File: apps/laboratory/services/deploy_status.py
from __future__ import annotations

from typing import Any

from django.conf import settings

from django.utils import timezone

from apps.laboratory.deploy_phases import deploy_phase_message
from apps.laboratory.models import StudentDeployment


def checker_public_url() -> str:
    return getattr(settings, 'CHECKER_PUBLIC_URL', 'https://checker.webflare.ru').rstrip('/')


def _snapshot_value(snapshot: dict[str, Any], key: str, default: Any = None) -> Any:
    value = snapshot.get(key, default)
    return default if value is None else value


def build_deployment_api_payload(deployment: StudentDeployment) -> dict[str, Any]:
    student = deployment.student
    snapshot = deployment.deploy_snapshot or {}
    phase = deployment.status
    label = _snapshot_value(snapshot, 'label') or deploy_phase_message(phase)
    legacy_status = {
        'queued': 'pending',
        'downloading': 'pending',
        'deploying': 'pending',
        'ready_to_test': 'running',
        'error': 'error',
    }.get(phase, 'pending')

    public_urls = deployment.access_urls or []
    url = deployment.deploy_url or _snapshot_value(snapshot, 'url')
    links = _snapshot_value(snapshot, 'links') or {}
    if not links:
        submission_uuid = deployment.last_submission_uuid
        links = {
            'docs': f'{checker_public_url()}/docs/',
            'status_api': f'{checker_public_url()}/api/deploy/status/',
        }
        if submission_uuid:
            links['panel'] = (
                f'{checker_public_url()}/deploy/panel/'
                f'?submission_uuid={submission_uuid}&embed=1'
            )

    return {
        'student_id': student.id,
        'student_username': student.username,
        'phase': phase,
        'status': legacy_status,
        'label': label,
        'status_label': label,
        'hint': _snapshot_value(snapshot, 'hint', ''),
        'progress': _snapshot_value(snapshot, 'progress'),
        'message': _snapshot_value(snapshot, 'message') or label,
        'url': url or None,
        'access_urls': public_urls,
        'public_urls': public_urls,
        'error': _snapshot_value(snapshot, 'error'),
        'links': links,
        'steps': _snapshot_value(snapshot, 'steps') or [],
        'messages': _snapshot_value(snapshot, 'messages') or [],
        'last_submission_uuid': str(deployment.last_submission_uuid) if deployment.last_submission_uuid else None,
        'submission_uuid': str(deployment.last_submission_uuid) if deployment.last_submission_uuid else None,
        'checker_project_id': deployment.checker_project_id,
        'public_base_url': deployment.public_base_url or '',
        'updated_at': deployment.updated_at.isoformat() if deployment.updated_at else None,
        'traceback': deployment.traceback or '',
    }


def _user_facing_error(data: dict[str, Any]) -> str:
    summary = (data.get('error_summary') or '').strip()
    if summary:
        return summary[:500]
    err = (data.get('error') or '').strip()
    if not err:
        return ''
    if len(err) > 500 or 'HTTPConnectionPool' in err or 'Traceback' in err:
        return 'Не удалось опубликовать проект на тестовом стенде'
    return err[:500]


def _append_deploy_messages(existing: list, data: dict[str, Any]) -> list:
    messages = list(existing or [])
    seen = {m.get('text') for m in messages if isinstance(m, dict) and m.get('text')}
    phase = data.get('phase') or data.get('status')
    at = data.get('updated_at') or timezone.now().isoformat()

    for field, kind in (('label', 'status'), ('hint', 'hint'), ('message', 'status')):
        text = data.get(field)
        if not text:
            continue
        text = str(text).strip()
        if not text or text in seen:
            continue
        messages.append({'text': text, 'phase': phase, 'kind': kind, 'at': at})
        seen.add(text)

    err = _user_facing_error(data)
    if err and err not in seen:
        messages.append({'text': err, 'phase': 'error', 'kind': 'error', 'at': at})

    return messages[-40:]


def apply_checker_payload_to_deployment(deployment: StudentDeployment, data: dict[str, Any]) -> None:
    from apps.laboratory.deploy_phases import normalize_deploy_phase

    raw_phase = data.get('phase') or data.get('status', 'error')
    deployment.status = normalize_deploy_phase(raw_phase)
    deployment.access_urls = data.get('public_urls') or data.get('access_urls') or []
    deployment.deploy_url = data.get('url') or ''
    deployment.public_base_url = data.get('public_base_url') or deployment.public_base_url or ''
    if data.get('submission_uuid'):
        deployment.last_submission_uuid = data['submission_uuid']
    deployment.checker_project_id = data.get('checker_project_id')

    user_error = _user_facing_error(data)
    prev_snapshot = deployment.deploy_snapshot or {}
    deployment.deploy_snapshot = {
        key: data[key]
        for key in ('label', 'hint', 'progress', 'message', 'url', 'links', 'steps')
        if data.get(key) is not None
    }
    deployment.deploy_snapshot['messages'] = _append_deploy_messages(
        prev_snapshot.get('messages') or data.get('messages'),
        data,
    )
    if user_error:
        deployment.deploy_snapshot['error'] = user_error

File: frontend/src/lib/deployStatus.ts
import type { DeploymentInfo } from '../components/DeploymentStatus'

export type DeployStep = {
  phase: string
  label: string
  done: boolean
  active: boolean
}

export type DeployLinks = {
  docs?: string
  panel?: string
  status_api?: string
}

export type DeployFeedMessage = {
  text: string
  phase?: string
  kind?: 'status' | 'hint' | 'error' | 'success'
  at?: string
}

export type CheckerDeployStatus = {
  submission_uuid?: string
  student_id?: number
  phase?: string
  status?: string
  label?: string
  hint?: string
  progress?: number
  message?: string
  url?: string | null
  public_urls?: DeploymentInfo['access_urls']
  access_urls?: DeploymentInfo['access_urls']
  error?: string | null
  links?: DeployLinks
  steps?: DeployStep[]
  messages?: DeployFeedMessage[]
  updated_at?: string
  checker_project_id?: number | null
  last_submission_uuid?: string | null
  status_label?: string
  traceback?: string
}

const IN_PROGRESS_PHASES = new Set(['queued', 'downloading', 'deploying', 'pending'])

const DEFAULT_STEPS: DeployStep[] = [
  { phase: 'queued', label: 'Подготовка', done: false, active: true },
  { phase: 'downloading', label: 'Загрузка проекта', done: false, active: false },
  { phase: 'deploying', label: 'Развёртывание', done: false, active: false },
  { phase: 'ready_to_test', label: 'Готово', done: false, active: false },
]

const PHASE_PROGRESS: Record<string, number> = {
  queued: 12,
  pending: 12,
  downloading: 38,
  deploying: 72,
  ready_to_test: 100,
  running: 100,
  error: 100,
}

function resolvePhase(raw: CheckerDeployStatus): string {
  if (raw.phase) return raw.phase
  if (raw.status === 'running') return 'ready_to_test'
  if (raw.status === 'pending') return 'queued'
  if (raw.status === 'error') return 'error'
  return raw.status || 'queued'
}

export function sanitizeDeployError(raw: string): string {
  const trimmed = raw.trim()
  if (!trimmed) return ''

  if (
    trimmed.includes('HTTPConnectionPool')
    || trimmed.includes('Traceback')
    || trimmed.includes('Connection refused')
    || trimmed.includes('NewConnectionError')
  ) {
    const line = trimmed
      .split('\n')
      .map((l) => l.trim())
      .find((l) => l && !l.startsWith('File ') && !l.startsWith('raise '))
    if (line && line.length <= 180) return line
    return 'Не удалось опубликовать проект на тестовом стенде'
  }

  if (trimmed.length > 280) return `${trimmed.slice(0, 277)}…`
  return trimmed
}

export function normalizeDeploymentInfo(raw: CheckerDeployStatus | DeploymentInfo): DeploymentInfo {
  const checker = raw as CheckerDeployStatus
  const phase = resolvePhase(checker)
  const access_urls = checker.public_urls ?? raw.access_urls ?? []
  const label = raw.label ?? raw.status_label
  const links = raw.links

  return {
    status: phase,
    phase,
    label,
    status_label: label,
    hint: raw.hint,
    progress: typeof raw.progress === 'number' ? raw.progress : PHASE_PROGRESS[phase],
    url: raw.url ?? null,
    access_urls,
    error: raw.error ? sanitizeDeployError(String(raw.error)) : null,
    links,
    steps: raw.steps?.length ? raw.steps : undefined,
    messages: raw.messages?.length ? raw.messages : undefined,
    last_submission_uuid: raw.last_submission_uuid ?? checker.submission_uuid ?? null,
    updated_at: raw.updated_at,
    checker_project_id: raw.checker_project_id,
    traceback: raw.traceback,
  }
}

export function deploymentDocsUrl(deployment?: DeploymentInfo | null): string {
  return (
    deployment?.links?.docs ||
    (import.meta.env.VITE_DEPLOY_DOCS_URL as string | undefined) ||
    'https://checker.webflare.ru/docs/'
  )
}

export function isDeploymentInProgress(deployment?: DeploymentInfo | null): boolean {
  if (!deployment) return false
  const phase = deployment.phase ?? deployment.status
  return IN_PROGRESS_PHASES.has(phase)
}

export function deploymentMatchesSubmission(
  deployment: DeploymentInfo,
  submissionUuid?: string | null,
) {
  if (!deployment.last_submission_uuid || !submissionUuid) return true
  return deployment.last_submission_uuid === submissionUuid
}

export function mergeDeploymentInfo(
  primary: DeploymentInfo | null | undefined,
  secondary: DeploymentInfo | null | undefined,
): DeploymentInfo | null {
  if (!primary) return secondary ?? null
  if (!secondary) return primary

  const primaryTime = primary.updated_at ? Date.parse(primary.updated_at) : 0
  const secondaryTime = secondary.updated_at ? Date.parse(secondary.updated_at) : 0
  const newer = secondaryTime >= primaryTime ? secondary : primary
  const older = newer === secondary ? primary : secondary

  return normalizeDeploymentInfo({
    ...older,
    ...newer,
    access_urls: newer.access_urls?.length ? newer.access_urls : older.access_urls,
    steps: newer.steps?.length ? newer.steps : older.steps,
    messages: (newer.messages?.length ? newer.messages : older.messages),
    links: { ...older.links, ...newer.links },
  })
}

export function buildFallbackSteps(deployment: DeploymentInfo): DeployStep[] {
  if (deployment.steps?.length) return deployment.steps

  const phase = deployment.phase ?? deployment.status
  const order = ['queued', 'downloading', 'deploying', 'ready_to_test']
  const activeIndex = phase === 'error'
    ? order.indexOf('deploying')
    : phase === 'ready_to_test' || phase === 'running'
      ? order.length
      : Math.max(0, order.indexOf(phase))

  return DEFAULT_STEPS.map((step, index) => {
    const done = phase === 'ready_to_test' || phase === 'running'
      ? true
      : index < activeIndex
    const active = phase === 'error'
      ? index === activeIndex
      : phase !== 'ready_to_test' && phase !== 'running' && index === activeIndex
    return { ...step, done, active }
  })
}

export function checkerPanelEmbedUrl(submissionUuid: string): string {
  const checker =
    (import.meta.env.VITE_CHECKER_PUBLIC_URL as string | undefined) ||
    'https://checker.webflare.ru'
  return `${checker.replace(/\/$/, '')}/deploy/panel/?submission_uuid=${encodeURIComponent(submissionUuid)}&embed=1`
}
